src/plan_add_hashagg.c

```diff
@@ -0,0 +1,595 @@
+
+
+	HASHCTL		hashctl = {
+		.keysize = sizeof(Oid),
+		.entrysize = sizeof(CustomEstimateForFunctionInfoHashEntry)
+	};
+
+	custom_estimate_func_hash = hash_create("custom_estimate_func_hash", _MAX_HASHAGG_FUNCTIONS, &hashctl, HASH_ELEM | HASH_BLOBS);
+
+	for (i = 0; i < _MAX_HASHAGG_FUNCTIONS; i++)
+	{
+		CustomEstimateForFunctionInfo def = custom_estimate_func_info[i];
+		CustomEstimateForFunctionInfoHashEntry *hash_entry;
+		bool		function_found = false;
+		bool		hash_found;
+		FuncCandidateList funclist;
+
+		if (def.extension_function)
+			funclist = FuncnameGetCandidates(list_make2(makeString(extension_schema_name), makeString(def.function_name)),
+											 def.nargs, NIL, false, false, false);
+		else
+			funclist = FuncnameGetCandidates(list_make1(makeString(def.function_name)),
+											 def.nargs, NIL, false, false, false);
+
+		/* check types */
+		while (!function_found && funclist != NULL)
+		{
+			if (funclist->nargs != def.nargs || !function_types_equal(funclist->args, def.arg_types, def.nargs))
+				funclist = funclist->next;
+			else
+				function_found = true;
+		}
+
+		if (!function_found)
+			elog(ERROR, "oid lookup failed for the function %s with %d args",
+				 def.function_name, def.nargs);
+
+		custom_estimate_func_info[i].function_oid = funclist->oid;
+
+		hash_entry = hash_search(custom_estimate_func_hash, &(custom_estimate_func_info[i].function_oid), HASH_ENTER, &hash_found);
+		Assert(!hash_found);
+		hash_entry->entry = &custom_estimate_func_info[i];
+	}
+}
+
+static CustomEstimateForFunctionInfo *
+get_custom_estimate_func_info(Oid function_oid)
+{
+	CustomEstimateForFunctionInfoHashEntry *hash_entry;
+
+	if (NULL == custom_estimate_func_hash)
+		initialize_custom_estimate_func_info();
+
+	hash_entry = hash_search(custom_estimate_func_hash, &function_oid, HASH_FIND, NULL);
+	return (hash_entry != NULL ? hash_entry->entry : NULL);
+}
+
+
+static double estimate_max_spread_expr(PlannerInfo *root, Expr *expr);
+
+/* Estimate the max spread on a time var in terms of the internal time representation.
+ * Note that this will happen on the hypertable var in most cases. Therefore this is
+ * a huge overestimate in many cases where there is a WHERE clause on time.
+ */
+static double
+estimate_max_spread_var(PlannerInfo *root, Var *var)
+{
+	VariableStatData vardata;
+	Oid			ltop;
+	Datum		max_datum,
+				min_datum;
+	int64		max,
+				min;
+	bool		valid;
+
+	examine_variable(root, (Node *) var, 0, &vardata);
+	get_sort_group_operators(var->vartype, true, false, false, &ltop, NULL, NULL, NULL);
+	valid = get_variable_range(root, &vardata, ltop, &min_datum, &max_datum);
+	ReleaseVariableStats(vardata);
+
+	if (!valid)
+		return INVALID_ESTIMATE;
+
+	max = time_value_to_internal(max_datum, var->vartype, true);
+	min = time_value_to_internal(min_datum, var->vartype, true);
+
+	if (max < 0 || min < 0)
+		return INVALID_ESTIMATE;
+
+	return ((double) max - min);
```

Comment:
Minor nit, and not sure it matters. But strictly, you might want to cast after the computation. I believe this casts `max` and then deducts `min` as an integer.

We might want:

```
return (double) (max - min);
```